pageserver/src/restore_local_repo.rs

```diff
@@ -415,26 +415,37 @@ pub fn save_decoded_record(
         }
     } else if decoded.xl_rmid == pg_constants::RM_XACT_ID {
         let info = decoded.xl_info & pg_constants::XLOG_XACT_OPMASK;
-        if info == pg_constants::XLOG_XACT_COMMIT
-            || info == pg_constants::XLOG_XACT_COMMIT_PREPARED
-            || info == pg_constants::XLOG_XACT_ABORT
+        if info == pg_constants::XLOG_XACT_COMMIT || info == pg_constants::XLOG_XACT_ABORT {
+            let parsed_xact = XlXactParsedRecord::decode(&mut buf, decoded.xl_xid, decoded.xl_info);
+            save_xact_record(timeline, lsn, &parsed_xact, decoded)?;
+        } else if info == pg_constants::XLOG_XACT_COMMIT_PREPARED
             || info == pg_constants::XLOG_XACT_ABORT_PREPARED
         {
             let parsed_xact = XlXactParsedRecord::decode(&mut buf, decoded.xl_xid, decoded.xl_info);
             save_xact_record(timeline, lsn, &parsed_xact, decoded)?;
-        } else if info == pg_constants::XLOG_XACT_PREPARE {
-            let rec = WALRecord {
+            // Remove twophase file. see RemoveTwoPhaseFile() in postgres code
+            info!(
+                "unlink twophaseFile for xid {} parsed_xact.xid {} here",
+                decoded.xl_xid, parsed_xact.xid
+            );
+            timeline.put_unlink(
+                RelishTag::TwoPhase {
+                    xid: decoded.xl_xid,
```

Comment:
This should use `parsed_xact.xid`, not `decoded.xl_xid`. The `xl_xid` in the header of a commit/abort prepared record doesn't mean anything useful, and is usually just 0. The xid in the record payload is what matters.

(I was hacking on the layered-repo branch, and happened to add an assertion in `put_unlink` there to assert that put_unlink is never called for a relish that has already been unlinked, and it tripped because of this. In the tests, two twophase transactions were prepared and later committed/aborted, but because of this bug, both lead to `put_unlink(xid: 0)` call, and the second one tripped the assertion.)

I'm surprised that this didn't cause the `test_twophase` test to fail even on 'main'.

Reply:
Ah, Postgres checks the CLOG for each prepared transaction at startup and ignores twophase states files for XIDs that are already marked as committed or aborted in the CLOG. But it does print a warning about them:

```
2021-08-12 21:37:49.380 GMT [1257097] LOG:  starting PostgreSQL 14beta1 on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit
2021-08-12 21:37:49.380 GMT [1257097] LOG:  listening on IPv4 address "127.0.0.1", port 55433
2021-08-12 21:37:49.380 GMT [1257097] LOG:  listening on Unix socket "/tmp/.s.PGSQL.55433"
2021-08-12 21:37:49.382 GMT [1257098] LOG:  database system was shut down at 2021-08-12 21:37:46 GMT
2021-08-12 21:37:49.382 GMT [1257098] WARNING:  removing stale two-phase state file for transaction 737
2021-08-12 21:37:49.382 GMT [1257098] LOG:  Continue writing WAL at 0/16DEAF0
2021-08-12 21:37:49.382 GMT [1257098] LOG:  recovering prepared transaction 736 from shared memory
2021-08-12 21:37:49.382 GMT [1257098] LOG:  recovering prepared transaction 735 from shared memory
```

Reply:
Opened https://github.com/zenithdb/zenith/issues/422 for this